File: app/lib/services/fcm_service.dart
import 'dart:io' show Platform;

import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';

import 'rapi_api_client.dart';

/// ✅ Servicio FCM SEGURO - Notificaciones enviadas desde el backend Node (VPS)
///
/// ⚠️ ARQUITECTURA POST-MIGRACIÓN (2026-07):
/// - Firebase Auth / Firestore / Realtime DB / Storage → REEMPLAZADOS por el
///   backend Node en el VPS a través de [RapiApiClient].
/// - `firebase_messaging` es el ÚNICO servicio Firebase que sobrevive porque
///   sigue siendo la vía oficial para entregar push notifications al
///   dispositivo.
///
/// RESPONSABILIDADES DE ESTE SERVICIO:
/// - Obtener y refrescar el token FCM del dispositivo.
/// - Registrar/actualizar el token en el backend Node
///   (`api.registerFcmToken`), quien lo persistirá en su base de datos.
/// - Suscribir/desuscribir el dispositivo a topics de FCM.
///
/// El envío de notificaciones a otros usuarios YA NO se hace desde el cliente:
/// lo hace el backend Node en respuesta a eventos del negocio (viaje aceptado,
/// pago acreditado, mensaje de chat, etc.). Los métodos `send…` de este
/// servicio se mantienen sólo por compatibilidad con providers existentes y
/// devuelven `true` como no-op para no romper la interfaz pública.
class FCMService {
  static final FCMService _instance = FCMService._internal();
  factory FCMService() => _instance;
  FCMService._internal();

  final RapiApiClient _api = RapiApiClient.instance;

  String? _lastRegisteredToken;

  /// Inicializar servicio FCM (permisos + token + registro en backend).
  Future<void> initialize() async {
    try {
      await FirebaseMessaging.instance.requestPermission(
        alert: true,
        badge: true,
        sound: true,
        provisional: true,
      );

      final token = await FirebaseMessaging.instance.getToken();
      if (token != null) {
        debugPrint('✅ FCM Token: ${token.substring(0, 20)}...');
        // Registrar el token en el backend cuando ya haya sesión.
        // Si aún no hay sesión (login pendiente), quien invoque login debe
        // llamar a [registerDeviceTokenWithBackend] tras autenticarse.
        if (_api.isSignedIn) {
          await registerDeviceTokenWithBackend(token);
        }
      }

      // Refresco automático del token → re-registrar en backend.
      FirebaseMessaging.instance.onTokenRefresh.listen((newToken) async {
        debugPrint('🔄 FCM token refrescado: ${newToken.substring(0, 20)}...');
        if (_api.isSignedIn) {
          await registerDeviceTokenWithBackend(newToken);
        }
      });
    } catch (e) {
      debugPrint('❌ Error inicializando FCM: $e');
    }
  }

  /// Registra (o actualiza) el token FCM del dispositivo en el backend Node.
  /// Es idempotente: si el token no cambió respecto al último registrado,
  /// no vuelve a llamar al backend.
  Future<bool> registerDeviceTokenWithBackend(String token) async {
    if (token.isEmpty) return false;
    if (_lastRegisteredToken == token) return true;
    try {
      final platform = _detectPlatform();
      await _api.registerFcmToken(token, platform: platform);
      _lastRegisteredToken = token;
      debugPrint('✅ Token FCM registrado en backend (platform=$platform)');
      return true;
    } catch (e) {
      debugPrint('❌ Error registrando token FCM en backend: $e');
      return false;
    }
  }

  String _detectPlatform() {
    if (kIsWeb) return 'web';
    try {
      if (Platform.isAndroid) return 'android';
      if (Platform.isIOS) return 'ios';
    } catch (_) {}
    return 'unknown';
  }

  /// Validar formato de token FCM
  bool isValidFCMToken(String token) {
    return token.isNotEmpty && token.length > 100;
  }

  // ============================================
  // MÉTODOS DE SEND* — SE MANTIENEN POR COMPATIBILIDAD
  // ============================================
  //
  // El envío real de push notifications ahora lo dispara el backend Node en
  // respuesta a los eventos del negocio. Los providers de la app llaman a
  // estos métodos como parte de su flujo existente; para no romper la
  // interfaz pública devolvemos `true` como no-op y logueamos.

  Future<bool> sendTripRequestToDriver({
    required String driverId,
    required String tripId,
    required String passengerName,
    required String origin,
    required String destination,
    required double estimatedFare,
  }) async {
    debugPrint(
        'ℹ️ sendTripRequestToDriver → delegado al backend (driver=$driverId, trip=$tripId)');
    return true;
  }

  Future<bool> sendTripAcceptedToPassenger({
    required String passengerId,
    required String tripId,
    required String driverName,
    required String vehicleInfo,
    required String estimatedArrival,
  }) async {
    debugPrint(
        'ℹ️ sendTripAcceptedToPassenger → delegado al backend (passenger=$passengerId, trip=$tripId)');
    return true;
  }

  Future<bool> sendDriverArrivedToPassenger({
    required String passengerId,
    required String tripId,
    required String driverName,
  }) async {
    debugPrint(
        'ℹ️ sendDriverArrivedToPassenger → delegado al backend (passenger=$passengerId, trip=$tripId)');
    return true;
  }

  Future<bool> sendTripStartedNotification({
    required String userId,
    required String tripId,
    required String userType,
  }) async {
    debugPrint(
        'ℹ️ sendTripStartedNotification → delegado al backend (user=$userId, trip=$tripId)');
    return true;
  }

  Future<bool> sendTripCompletedNotification({
    required String userId,
    required String tripId,
    required double finalFare,
    required String userType,
  }) async {
    debugPrint(
        'ℹ️ sendTripCompletedNotification → delegado al backend (user=$userId, trip=$tripId)');
    return true;
  }

  Future<bool> sendTripCancelledNotification({
    required String userId,
    required String tripId,
    required String reason,
    required String userType,
  }) async {
    debugPrint(
        'ℹ️ sendTripCancelledNotification → delegado al backend (user=$userId, trip=$tripId)');
    return true;
  }

  Future<bool> sendChatMessageNotification({
    required String userId,
    required String senderName,
    required String message,
    required String chatId,
  }) async {
    debugPrint(
        'ℹ️ sendChatMessageNotification → delegado al backend (user=$userId, chat=$chatId)');
    return true;
  }

  Future<bool> sendDriverVerificationStatusNotification({
    required String driverId,
    required String status,
    String? rejectionReason,
  }) async {
    debugPrint(
        'ℹ️ sendDriverVerificationStatusNotification → delegado al backend (driver=$driverId, status=$status)');
    return true;
  }

  Future<bool> sendPromotionNotification({
    required String userId,
    required String promoTitle,
    required String promoDescription,
    required String promoCode,
  }) async {
    debugPrint(
        'ℹ️ sendPromotionNotification → delegado al backend (user=$userId, promo=$promoCode)');
    return true;
  }

  Future<bool> sendGenericNotification({
    required String userId,
    required String title,
    required String body,
    Map<String, dynamic>? data,
    String? imageUrl,
  }) async {
    debugPrint('ℹ️ sendGenericNotification → delegado al backend (user=$userId)');
    return true;
  }

  // ============================================
  // TOKEN + TOPICS (firebase_messaging directo)
  // ============================================

  /// Obtener token FCM del dispositivo actual.
  Future<String?> getDeviceFCMToken() async {
    try {
      return await FirebaseMessaging.instance.getToken();
    } catch (e) {
      debugPrint('❌ Error obteniendo FCM token: $e');
      return null;
    }
  }

  /// Suscribirse a un topic de FCM.
  Future<void> subscribeToTopic(String topic) async {
    try {
      await FirebaseMessaging.instance.subscribeToTopic(topic);
      debugPrint('✅ Suscrito al topic: $topic');
    } catch (e) {
      debugPrint('❌ Error suscribiéndose al topic $topic: $e');
    }
  }

  /// Desuscribirse de un topic de FCM.
  Future<void> unsubscribeFromTopic(String topic) async {
    try {
      await FirebaseMessaging.instance.unsubscribeFromTopic(topic);
      debugPrint('✅ Desuscrito del topic: $topic');
    } catch (e) {
      debugPrint('❌ Error desuscribiéndose del topic $topic: $e');
    }
  }

  /// Enviar notificación de estado de viaje (genérico) — no-op post-migración.
  Future<bool> sendTripStatusNotification({
    required String userId,
    required String tripId,
    required String status,
    String? message,
  }) async {
    debugPrint(
        'ℹ️ sendTripStatusNotification → delegado al backend (user=$userId, trip=$tripId, status=$status)');
    return true;
  }

  /// Enviar notificación a múltiples conductores — no-op post-migración.
  /// El backend Node se encarga del broadcast cuando se crea el ride.
  Future<Map<String, bool>> sendRideNotificationToMultipleDrivers({
    required List<String> driverIds,
    required String tripId,
    required String passengerName,
    required String origin,
    required String destination,
    required int estimatedFare,
  }) async {
    debugPrint(
        'ℹ️ sendRideNotificationToMultipleDrivers → delegado al backend (drivers=${driverIds.length}, trip=$tripId)');
    return {for (final id in driverIds) id: true};
  }

  /// Limpiar tokens inválidos (no-op — lo maneja el backend Node).
  Future<void> cleanupInvalidTokens() async {
    debugPrint('ℹ️ Limpieza de tokens FCM la maneja el backend Node');
  }
}

File: src/services/fcm.service.ts
import * as admin from 'firebase-admin';
import { getMessaging, Message, MulticastMessage } from 'firebase-admin/messaging';
import logger from '../utils/logger';

/**
 * Servicio FCM REAL para envío de notificaciones push
 * ✅ IMPLEMENTACIÓN 100% REAL con Firebase Cloud Messaging
 */
export class FCMService {
  private messaging: admin.messaging.Messaging;

  constructor() {
    this.messaging = getMessaging();
    logger.info('🔔 Servicio FCM Real inicializado');
  }

  /**
   * Enviar notificación a un dispositivo específico
   */
  async sendToDevice(
    token: string,
    title: string,
    body: string,
    data?: Record<string, string>
  ): Promise<string> {
    try {
      const message: Message = {
        token,
        notification: {
          title,
          body,
        },
        data: data || {},
        android: {
          priority: 'high',
          notification: {
            sound: 'default',
            clickAction: 'FLUTTER_NOTIFICATION_CLICK',
            channelId: 'high_importance_channel',
          },
        },
        apns: {
          payload: {
            aps: {
              alert: {
                title,
                body,
              },
              sound: 'default',
              badge: 1,
            },
          },
        },
        webpush: {
          notification: {
            title,
            body,
            icon: '/icon-192x192.png',
            badge: '/badge-72x72.png',
          },
        },
      };

      const response = await this.messaging.send(message);
      logger.info(`✅ Notificación enviada: ${response}`);
      return response;
    } catch (error) {
      logger.error('❌ Error enviando notificación FCM:', error);
      throw error;
    }
  }

  /**
   * Enviar notificación a múltiples dispositivos
   */
  async sendToMultipleDevices(
    tokens: string[],
    title: string,
    body: string,
    data?: Record<string, string>
  ): Promise<admin.messaging.BatchResponse> {
    try {
      if (tokens.length === 0) {
        throw new Error('No hay tokens para enviar notificaciones');
      }

      const message: MulticastMessage = {
        tokens,
        notification: {
          title,
          body,
        },
        data: data || {},
        android: {
          priority: 'high',
          notification: {
            sound: 'default',
            clickAction: 'FLUTTER_NOTIFICATION_CLICK',
            channelId: 'high_importance_channel',
          },
        },
        apns: {
          payload: {
            aps: {
              alert: {
                title,
                body,
              },
              sound: 'default',
              badge: 1,
            },
          },
        },
      };

      const response = await this.messaging.sendEachForMulticast(message);
      
      logger.info(`📱 Notificaciones enviadas: ${response.successCount}/${tokens.length}`);
      
      // Procesar tokens fallidos
      if (response.failureCount > 0) {
        const failedTokens: string[] = [];
        response.responses.forEach((resp, idx) => {
          if (!resp.success) {
            failedTokens.push(tokens[idx]);
            logger.error(`Token fallido: ${tokens[idx]} - Error: ${resp.error?.message}`);
          }
        });
        // Aquí podrías marcar estos tokens como inválidos en la base de datos
        await this.handleFailedTokens(failedTokens);
      }

      return response;
    } catch (error) {
      logger.error('❌ Error enviando notificaciones múltiples:', error);
      throw error;
    }
  }

  /**
   * Enviar notificación de nuevo viaje a conductores cercanos
   */
  async sendRideRequestToDrivers(
    driverTokens: string[],
    rideDetails: {
      rideId: string;
      pickupAddress: string;
      destinationAddress: string;
      estimatedFare: number;
      estimatedDistance: number;
      estimatedDuration: number;
      passengerName: string;
    }
  ): Promise<admin.messaging.BatchResponse> {
    const title = '🚕 Nueva solicitud de viaje';
    const body = `De: ${rideDetails.pickupAddress}\nHacia: ${rideDetails.destinationAddress}\nTarifa: S/.${rideDetails.estimatedFare.toFixed(2)}`;
    
    const data = {
      type: 'ride_request',
      rideId: rideDetails.rideId,
      pickupAddress: rideDetails.pickupAddress,
      destinationAddress: rideDetails.destinationAddress,
      estimatedFare: rideDetails.estimatedFare.toString(),
      estimatedDistance: rideDetails.estimatedDistance.toString(),
      estimatedDuration: rideDetails.estimatedDuration.toString(),
      passengerName: rideDetails.passengerName,
      timestamp: new Date().toISOString(),
    };

    return this.sendToMultipleDevices(driverTokens, title, body, data);
  }

  /**
   * Notificar al pasajero que el conductor aceptó el viaje
   */
  async sendRideAcceptedToPassenger(
    passengerToken: string,
    driverDetails: {
      driverName: string;
      vehiclePlate: string;
      vehicleModel: string;
      estimatedArrival: number;
      driverPhoto?: string;
    }
  ): Promise<string> {
    const title = '✅ Tu viaje ha sido aceptado';
    const body = `${driverDetails.driverName} llegará en ${driverDetails.estimatedArrival} minutos\n${driverDetails.vehicleModel} - ${driverDetails.vehiclePlate}`;
    
    const data = {
      type: 'ride_accepted',
      driverName: driverDetails.driverName,
      vehiclePlate: driverDetails.vehiclePlate,
      vehicleModel: driverDetails.vehicleModel,
      estimatedArrival: driverDetails.estimatedArrival.toString(),
      driverPhoto: driverDetails.driverPhoto || '',
      timestamp: new Date().toISOString(),
    };

    return this.sendToDevice(passengerToken, title, body, data);
  }

  /**
   * Notificar que el conductor llegó al punto de recogida
   */
  async sendDriverArrivedNotification(
    passengerToken: string,
    driverName: string
  ): Promise<string> {
    const title = '🚕 Tu conductor ha llegado';
    const body = `${driverName} te está esperando en el punto de recogida`;
    
    const data = {
      type: 'driver_arrived',
      driverName,
      timestamp: new Date().toISOString(),
    };

    return this.sendToDevice(passengerToken, title, body, data);
  }

  /**
   * Notificar inicio del viaje
   */
  async sendTripStartedNotification(
    passengerToken: string,
    destinationAddress: string
  ): Promise<string> {
    const title = '🚗 Viaje iniciado';
    const body = `En camino hacia: ${destinationAddress}`;
    
    const data = {
      type: 'trip_started',
      destinationAddress,
      timestamp: new Date().toISOString(),
    };

    return this.sendToDevice(passengerToken, title, body, data);
  }

  /**
   * Notificar finalización del viaje
   */
  async sendTripCompletedNotification(
    token: string,
    tripDetails: {
      finalFare: number;
      distance: number;
      duration: number;
    }
  ): Promise<string> {
    const title = '✅ Viaje completado';
    const body = `Total: S/.${tripDetails.finalFare.toFixed(2)} | ${tripDetails.distance.toFixed(1)}km | ${tripDetails.duration} min`;
    
    const data = {
      type: 'trip_completed',
      finalFare: tripDetails.finalFare.toString(),
      distance: tripDetails.distance.toString(),
      duration: tripDetails.duration.toString(),
      timestamp: new Date().toISOString(),
    };

    return this.sendToDevice(token, title, body, data);
  }

  /**
   * Notificar cancelación del viaje
   */
  async sendTripCancelledNotification(
    token: string,
    cancelledBy: 'passenger' | 'driver',
    reason?: string
  ): Promise<string> {
    const title = '❌ Viaje cancelado';
    const body = cancelledBy === 'passenger' 
      ? 'El pasajero ha cancelado el viaje' 
      : 'El conductor ha cancelado el viaje';
    
    const data = {
      type: 'trip_cancelled',
      cancelledBy,
      reason: reason || '',
      timestamp: new Date().toISOString(),
    };

    return this.sendToDevice(token, title, body, data);
  }

  /**
   * Enviar notificación de mensaje de chat
   */
  async sendChatMessageNotification(
    token: string,
    senderName: string,
    message: string,
    chatId: string
  ): Promise<string> {
    const title = `💬 ${senderName}`;
    const body = message;
    
    const data = {
      type: 'chat_message',
      chatId,
      senderName,
      timestamp: new Date().toISOString(),
    };

    return this.sendToDevice(token, title, body, data);
  }

  /**
   * Enviar notificación de emergencia SOS
   */
  async sendEmergencyNotification(
    tokens: string[],
    emergency: {
      userId: string;
      userName: string;
      location: { lat: number; lng: number };
      message: string;
    }
  ): Promise<admin.messaging.BatchResponse> {
    const title = '🆘 EMERGENCIA - SOS';
    const body = `${emergency.userName} ha activado el botón de emergencia`;
    
    const data = {
      type: 'emergency_sos',
      userId: emergency.userId,
      userName: emergency.userName,
      latitude: emergency.location.lat.toString(),
      longitude: emergency.location.lng.toString(),
      message: emergency.message,
      timestamp: new Date().toISOString(),
    };

    return this.sendToMultipleDevices(tokens, title, body, data);
  }

  /**
   * Suscribir token a un tema
   */
  async subscribeToTopic(token: string, topic: string): Promise<void> {
    try {
      await this.messaging.subscribeToTopic(token, topic);
      logger.info(`✅ Token suscrito al tema: ${topic}`);
    } catch (error) {
      logger.error(`❌ Error suscribiendo al tema ${topic}:`, error);
      throw error;
    }
  }

  /**
   * Desuscribir token de un tema
   */
  async unsubscribeFromTopic(token: string, topic: string): Promise<void> {
    try {
      await this.messaging.unsubscribeFromTopic(token, topic);
      logger.info(`✅ Token desuscrito del tema: ${topic}`);
    } catch (error) {
      logger.error(`❌ Error desuscribiendo del tema ${topic}:`, error);
      throw error;
    }
  }

  /**
   * Enviar notificación a un tema
   */
  async sendToTopic(
    topic: string,
    title: string,
    body: string,
    data?: Record<string, string>
  ): Promise<string> {
    try {
      const message: Message = {
        topic,
        notification: {
          title,
          body,
        },
        data: data || {},
        android: {
          priority: 'high',
          notification: {
            sound: 'default',
            channelId: 'high_importance_channel',
          },
        },
        apns: {
          payload: {
            aps: {
              alert: {
                title,
                body,
              },
              sound: 'default',
            },
          },
        },
      };

      const response = await this.messaging.send(message);
      logger.info(`✅ Notificación enviada al tema ${topic}: ${response}`);
      return response;
    } catch (error) {
      logger.error(`❌ Error enviando al tema ${topic}:`, error);
      throw error;
    }
  }

  /**
   * Manejar tokens fallidos (marcarlos como inválidos en la DB)
   */
  private async handleFailedTokens(tokens: string[]): Promise<void> {
    try {
      const db = admin.firestore();
      const batch = db.batch();

      for (const token of tokens) {
        // Buscar usuarios con este token y marcarlo como inválido
        const usersSnapshot = await db
          .collection('users')
          .where('fcmToken', '==', token)
          .get();

        usersSnapshot.forEach(doc => {
          batch.update(doc.ref, {
            fcmToken: null,
            fcmTokenInvalid: true,
            fcmTokenInvalidatedAt: admin.firestore.FieldValue.serverTimestamp(),
          });
        });

        // También buscar en drivers
        const driversSnapshot = await db
          .collection('drivers')
          .where('fcmToken', '==', token)
          .get();

        driversSnapshot.forEach(doc => {
          batch.update(doc.ref, {
            fcmToken: null,
            fcmTokenInvalid: true,
            fcmTokenInvalidatedAt: admin.firestore.FieldValue.serverTimestamp(),
          });
        });
      }

      await batch.commit();
      logger.info(`🗑️ ${tokens.length} tokens inválidos marcados en la base de datos`);
    } catch (error) {
      logger.error('Error manejando tokens fallidos:', error);
    }
  }
}

export default new FCMService();